File: computing/interview/design-patterns/creational/prototype/csharp/PrototypeDemo.cs
using System;
using System.Collections.Generic;

namespace Interview.DesignPatterns.Creational.Prototype
{
    public abstract class DiagramNode
    {
        protected DiagramNode(string label)
        {
            Label = label;
        }

        public string Label { get; }
        public abstract DiagramNode Clone();
        public abstract void Render();
    }

    public sealed class CircleNode : DiagramNode
    {
        private readonly double _radius;

        public CircleNode(string label, double radius) : base(label)
        {
            _radius = radius;
        }

        public override DiagramNode Clone() => new CircleNode(Label, _radius);

        public override void Render()
        {
            Console.WriteLine($"Circle({Label}, r={_radius})");
        }
    }

    public sealed class RectangleNode : DiagramNode
    {
        private readonly double _width;
        private readonly double _height;

        public RectangleNode(string label, double width, double height) : base(label)
        {
            _width = width;
            _height = height;
        }

        public override DiagramNode Clone() => new RectangleNode(Label, _width, _height);

        public override void Render()
        {
            Console.WriteLine($"Rectangle({Label}, {_width}x{_height})");
        }
    }

    public sealed class PrototypeRegistry
    {
        private readonly Dictionary<string, DiagramNode> _prototypes = new();

        public void Register(string name, DiagramNode prototype)
        {
            _prototypes[name] = prototype;
        }

        public DiagramNode Create(string name)
        {
            if (!_prototypes.TryGetValue(name, out var prototype))
            {
                throw new InvalidOperationException("Prototype not found");
            }

            return prototype.Clone();
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            var registry = new PrototypeRegistry();
            registry.Register("small_circle", new CircleNode("Small", 1.0));
            registry.Register("wide_rect", new RectangleNode("Wide", 3.0, 1.0));

            var node1 = registry.Create("small_circle");
            var node2 = registry.Create("wide_rect");

            node1.Render();
            node2.Render();
        }
    }
}

File: computing/interview/design-patterns/creational/prototype/java/PrototypeDemo.java
package interview.designpatterns.creational.prototype;

import java.util.HashMap;
import java.util.Map;

abstract class DiagramNode {
    private final String label;

    DiagramNode(String label) {
        this.label = label;
    }

    String label() {
        return label;
    }

    abstract DiagramNode cloneNode();
    abstract void render();
}

final class CircleNode extends DiagramNode {
    private final double radius;

    CircleNode(String label, double radius) {
        super(label);
        this.radius = radius;
    }

    @Override
    DiagramNode cloneNode() {
        return new CircleNode(label(), radius);
    }

    @Override
    void render() {
        System.out.printf("Circle(%s, r=%.1f)%n", label(), radius);
    }
}

final class RectangleNode extends DiagramNode {
    private final double width;
    private final double height;

    RectangleNode(String label, double width, double height) {
        super(label);
        this.width = width;
        this.height = height;
    }

    @Override
    DiagramNode cloneNode() {
        return new RectangleNode(label(), width, height);
    }

    @Override
    void render() {
        System.out.printf("Rectangle(%s, %.1fx%.1f)%n", label(), width, height);
    }
}

final class PrototypeRegistry {
    private final Map<String, DiagramNode> prototypes = new HashMap<>();

    void registerNode(String name, DiagramNode node) {
        prototypes.put(name, node);
    }

    DiagramNode create(String name) {
        DiagramNode prototype = prototypes.get(name);
        if (prototype == null) {
            throw new IllegalArgumentException("Prototype not found");
        }
        return prototype.cloneNode();
    }
}

public final class PrototypeDemo {
    private PrototypeDemo() {}

    public static void main(String[] args) {
        PrototypeRegistry registry = new PrototypeRegistry();
        registry.registerNode("small_circle", new CircleNode("Small", 1.0));
        registry.registerNode("wide_rect", new RectangleNode("Wide", 3.0, 1.0));

        registry.create("small_circle").render();
        registry.create("wide_rect").render();
    }
}
